### PR TITLE
Fixed revision count and dropped observers

### LINKED ISSUE #1236: Broken revision count
I don't understand the intended behavior of  [revision_observer.rb](https://github.com/openSUSE/osem/blob/master/app/models/revision_observer.rb) .

It seems like it should increment `revision` column on every change of `room` , `track` , `event` and `conference` itself. If that's the case then then the count increments for the changes in `conference` model only and needs to be fixed.

### PR DESCRIPTION
**Pull request by shlok007**
Closes #1236 .

This PR fixes the `revision` count in conference and also switches to use concerns instead of observers to calculate revision count .

### CODE REVIEW

**Comment by shlok007** [comment]
Since, random travis failures were fixed in #1246 . Can the travis build for this PR be restarted please? :slightly_smiling_face:

**Comment by differentreality** [comment]
What do we want to do about this?

**Comment by shlok007** [comment]
Are the changes okay? should I squash and rebase?

**Comment by differentreality** [comment]
@lagartoflojo was it you that would take a look at this?

**Comment by lagartoflojo** [comment]
@differentreality will do!

File: app/models/concerns/revision_count.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+module RevisionCount
+  extend ActiveSupport::Concern
+
+  included do
+    after_update :increment_revision
+  end
+
+  def increment_revision
+    conference = if is_a?(Room)
```
Comment (lagartoflojo) [inline app/models/concerns/revision_count.rb]: I reaaaaally don't like checking the type of the class to see what method to call.
Prefer [duck typing](https://robots.thoughtbot.com/back-to-basics-polymorphism-and-ruby#duck-typing)!

We can just set a "requirement" for the `RevisionCount` module: if a class includes this module, it has to implement the `conference` method.
So:
Room could have `has_one :conference, through: :venue`
Conference: `def conference; self; end`
Event: `has_one :conference, through: :program`
That way, the `increment_revision` method is reduced to:
```ruby
def increment_revision
  conference.increment!(:revision)
end
```

File: app/models/concerns/revision_count.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+                   venue.conference
+                 else
+                   is_a?(Conference) ? self : program.conference
+                 end
+    conference.reload.increment(:revision)
+    conference.update_column(:revision, conference.revision)
+  rescue
```
Comment (lagartoflojo) [inline app/models/concerns/revision_count.rb]: Why do we need to rescue?

File: db/migrate/20170108053041_add_default_to_revision_in_conference.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddDefaultToRevisionInConference < ActiveRecord::Migration
+  def change
+    change_column :conferences, :revision, :integer, default: 0
```
Comment (lagartoflojo) [inline db/migrate/20170108053041_add_default_to_revision_in_conference.rb]: I'd add: `null: false`

File: app/models/concerns/revision_count.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+require 'active_support/concern'
```
Comment (shlok007) [inline app/models/concerns/revision_count.rb]: I seems to work fine without it.

Comment (lagartoflojo) [inline app/models/concerns/revision_count.rb]: Let's remove it then :)

File: app/models/concerns/revision_count.rb (hunk 1 above)
Comment (lagartoflojo) [inline app/models/concerns/revision_count.rb]: Ok, so the problem is not with duck typing, which works fine!

The problem is with `#increment!`. Let's first consider that `#increment_revision` gets called on `after_update`. With that in mind, the problem is that `#increment!` ALSO triggers callbacks, so it calls `after_update` and thus `#increment_revision` AGAIN. And again. And again. Which is why you get "stack level too deep".

So to update the `revision`, we need to do it using a method that will NOT trigger callbacks.

File: db/schema.rb [hunk 5]
```diff
@@ -13,20 +13,17 @@
 
 ActiveRecord::Schema.define(version: 20170603095900) do
 
-  # These are extensions that must be enabled in order to support this database
-  enable_extension "plpgsql"
-
   create_table "ahoy_events", force: :cascade do |t|
-    t.integer  "visit_id"
+    t.uuid     "visit_id",   limit: 16
```
Comment (lagartoflojo) [inline db/schema.rb]: Hmm why are we seeing all these changes in schema.rb?

File: spec/models/conference_spec.rb [hunk 6]
```diff
@@ -1656,4 +1656,32 @@
       expect{ conference.save }.to change{ EventSchedule.count }.from(2).to(1)
     end
   end
+
+  describe 'revision count' do
```
Comment (lagartoflojo) [inline spec/models/conference_spec.rb]: `describe '#revision'`

File: spec/models/conference_spec.rb [hunk 7]
```diff
@@ -1656,4 +1656,32 @@
...
+
+    let(:track) { create(:track, program: subject.program) }
+    let(:event) { create(:event, program: subject.program, track: track) }
+    let(:venue) { create(:venue, conference: subject) }
+    let(:room)  { create(:room, venue: venue) }
+
+    context 'should increment' do
```
Comment (lagartoflojo) [inline spec/models/conference_spec.rb]: Contexts are used when you have different cases, such as:

```ruby
context 'when user is logged in' do #....
context 'when user is logged out' do #....
```

As such, it does not make much sense to use one single `context`, so in this case we can just remove it.

File: spec/models/conference_spec.rb [hunk 8]
```diff
@@ -1656,4 +1656,32 @@
...
+
+      it 'for change in conference' do
+        subject.title = 'changed'
+        expect{ subject.save }.to change { subject.revision }.by(1)
+      end
+      it 'for change in event' do
```
Comment (lagartoflojo) [inline spec/models/conference_spec.rb]: Please put a blank line between `it` blocks :)
(I wonder why we don't have a RuboCop rule for this...)

File: app/models/concerns/revision_count.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+module RevisionCount
+  extend ActiveSupport::Concern
+
+  included do
+    after_update :increment_revision
+  end
+
+  def increment_revision
+    conference.increment!(:revision)
```
Comment (lagartoflojo) [inline app/models/concerns/revision_count.rb]: Ok, so the problem is not with duck typing, which works fine!

The problem is with `#increment!`. Let's first consider that `#increment_revision` gets called on `after_update`. With that in mind, the problem is that `#increment!` ALSO triggers callbacks, so it calls `after_update` and thus `#increment_revision` AGAIN. And again. And again. Which is why you get "stack level too deep".

So to update the `revision`, we need to do it using a method that will NOT trigger callbacks.

File: db/schema.rb (hunk 5 above)
Comment (shlok007) [inline db/schema.rb]: seems like something messed up while rebasing the PR. I'll correct it.

File: app/models/concerns/revision_count.rb (hunk 9 above)
Comment (shlok007) [inline app/models/concerns/revision_count.rb]: Oh! The way it was before did not trigger the callbacks and that is why we didn't encounter the error at that time.
`update_attribute` or `update_column` will not trigger callbacks so we can maybe leave it the way it was before:
```ruby
conference.reload.increment(:revision)
conference.update_column(:revision, conference.revision)
```

**Comment by lagartoflojo** [comment]
I just realized that we have to remove this line:
https://github.com/openSUSE/osem/blob/master/config/application.rb#L28

**Comment by shlok007** [comment]
@lagartoflojo Do you mean this line which has been removed here: https://github.com/openSUSE/osem/pull/1239/files#diff-b1fe55db50c712fef0673345e5b9c0d9L28

**Comment by lagartoflojo** [comment]
Woop! I missed that 😳

**Pull request merged**